org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/SplitPackageBinding.java

```diff
@@ -47,7 +47,13 @@ public static PackageBinding combine(PackageBinding binding, PackageBinding prev
 			return previous;
 		if (binding.subsumes(previous))
 			return binding;
-		SplitPackageBinding split = new SplitPackageBinding(previous, primaryModule);
+
+		final SplitPackageBinding split;
+		if (previous instanceof SplitPackageBinding previousSplit && previousSplit.enclosingModule == primaryModule) {
+			split = previousSplit;
+		} else {
+			split = new SplitPackageBinding(previous, primaryModule);
+		}
 		split.add(binding);
```

Comment:
Attention: this modifies the previous PackageBinding. A caller would not expect that from the signature.

Reply:
Of cause, and still I somehow missed it :/
Thanks for pointing it out. This can be a problem. I will replace it with a different change soon which should fix my problem.

Reply:
Well, seem to be a bit more tricky than I thought to find get a different approach. Wouldn't mind if someone finds a different solution to the problem I described in #2646.